src/parser.rs

```diff
@@ -211,38 +213,27 @@ impl<'a> Parser<'a> {
         }
     }
 
-    fn parse_unary_plus(&mut self) -> Option<UnaryPlus> {
-        assert!(self.token.kind() == Kind::Plus);
-        self.read_token(); // consume `+`
-        match self.token.kind() {
-            Kind::Integer => self.parse_integer().map(UnaryPlus::Integer),
-            Kind::FloatingPoint => self.parse_float().map(UnaryPlus::FloatingPoint),
-            Kind::Identifier => Some(UnaryPlus::Identifier(self.parse_string())),
-            _ => {
-                self.errors.push(format!(
-                    "Parse error when parsing unary plus expression {:?}",
-                    self.token
-                ));
-                None
-            }
-        }
-    }
-
-    fn parse_unary_minus(&mut self) -> Option<UnaryMinus> {
-        assert!(self.token.kind() == Kind::Minus);
-        self.read_token(); // consume `-`
-        match self.token.kind() {
-            Kind::Integer => self.parse_integer().map(UnaryMinus::Integer),
-            Kind::FloatingPoint => self.parse_float().map(UnaryMinus::FloatingPoint),
-            Kind::Identifier => Some(UnaryMinus::Identifier(self.parse_string())),
+    fn parse_unary_op(&mut self) -> Option<UnaryOp> {
+        assert!(self.token.kind() == Kind::Plus || self.token.kind() == Kind::Minus);
+        let sign = if self.token.kind() == Kind::Plus {
```

Comment:
I wonder if this part should now be moved back into the main matcher in `parse_expression`, i.e. distinct matches for `+` and `-`, which both call this function and supply the appropriate `Kind::Plus` or `Kind::Minus` as an argument?

This would simplify the logic here, and would allow the matcher in `parse_expression` to handle these unary operators in the same way we might handle other unary operators in the future...

Reply:
addressed